File: src/main/java/pl/mroz/buddiesapi/infrastructure/database/account/AccountRepositoryImpl.java
package pl.mroz.buddiesapi.infrastructure.database.account;

import lombok.RequiredArgsConstructor;
import org.springframework.data.jpa.repository.JpaRepository;
import pl.mroz.buddiesapi.domain.account.Account;
import pl.mroz.buddiesapi.domain.account.AccountRepository;

import java.util.List;
import java.util.UUID;

@RequiredArgsConstructor
public class AccountRepositoryImpl implements AccountRepository {

    private final AccountJpaRepository jpaRepository;

    @Override
    public List<Account> findAll() {
        return jpaRepository.findAll().stream()
                .map(Account::fromDb)
                .toList();
    }

    @Override
    public Account getById(UUID accountUUID) {
        return Account.fromDb(jpaRepository.getByAccountId(accountUUID));
    }

    @Override
    public Account save(Account account) {
        var savedEntity = jpaRepository.save(AccountEntity.fromDomain(account));
        return Account.fromDb(savedEntity);
    }

    @Override
    public void delete(Account account) {
        jpaRepository.delete(AccountEntity.fromDomain(account));
    }

    interface AccountJpaRepository extends JpaRepository<AccountEntity, UUID> {

        AccountEntity getByAccountId(UUID accountId);
    }
}

File: src/main/kotlin/buddiesapi/infrastructure/database/account/AccountRepositoryImpl.kt
package buddiesapi.infrastructure.database.account

import buddiesapi.domain.account.Account
import buddiesapi.domain.account.AccountRepository
import buddiesapi.infrastructure.database.account.AccountEntity.Companion.fromDomain
import java.util.UUID
import org.springframework.data.jpa.repository.JpaRepository

class AccountRepositoryImpl(private val jpaRepository: AccountJpaRepository) : AccountRepository {

    override fun findAll(): List<Account> = jpaRepository.findAll().stream()
        .map { entity: AccountEntity? -> entity?.let { Account.fromDb(it) } }
        .toList()


    override fun getById(accountUUID: UUID): Account? =
        jpaRepository.getByAccountId(accountUUID)?.let {
            Account.fromDb(it)
        }

    override fun save(account: Account): Account {
        val savedEntity = jpaRepository.save(fromDomain(account))
        return Account.fromDb(savedEntity)
    }

    override fun delete(account: Account) =
        jpaRepository.delete(fromDomain(account))

    interface AccountJpaRepository :
        JpaRepository<AccountEntity, UUID> {
        fun getByAccountId(accountId: UUID): AccountEntity?
    }
}
